compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,379 @@
+                            block_else: None,
+                            block_exit: right,
+                        });
+                    } else if left == *else_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: None,
+                            block_else: Some(left),
+                            block_exit: right,
+                        });
+                    }
+                }
+            }
+        } else if right_successors_len == 1 && next_right == Some(left) {
+            // Right branch joins the right branch, it is a if/else statement with no then
+            // I am not sure whether this case can happen, but it is not difficult to handle it
+            let cost = block_cost(right, &function.dfg);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    if right == *else_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: None,
+                            block_else: Some(right),
+                            block_exit: right,
+                        });
+                    } else if right == *then_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: Some(right),
+                            block_else: None,
+                            block_exit: right,
+                        });
+                    }
+                }
+            }
+        }
+    }
+    // A conditional exit would have exactly 2 predecessors
+    result.filter(|result| cfg.predecessors(result.block_exit).len() == 2)
+}
+
+/// Computes a cost estimate of a basic block
+/// returns u32::MAX if the block has side-effect instructions
+fn block_cost(block: BasicBlockId, dfg: &DataFlowGraph) -> u32 {
+    let mut cost: u32 = 0;
+
+    for instruction in dfg[block].instructions() {
+        let instruction_cost = match &dfg[*instruction] {
+            Instruction::Binary(binary) => {
+                match binary.operator {
+                    BinaryOp::Add { unchecked }
+                    | BinaryOp::Sub { unchecked }
+                    | BinaryOp::Mul { unchecked } => if unchecked { 1 } else { return u32::MAX },
+                    BinaryOp::Div
+                    | BinaryOp::Mod => return u32::MAX,
+                    BinaryOp::Eq => 2,
+                    BinaryOp::Lt => 5,
+                    BinaryOp::And
+                    | BinaryOp::Or
+                    | BinaryOp::Xor => 1,   //todo
```

Comment:
This TODO doesn't have a description and there are a couple TODOs in this function. Could you make issues to handle them if we are not going to in this PR?

Reply:
Happy to address these in this PR. The point is to provide realistic brillig cost for each opcode. The idea is to improve execution speed so the cost should be the runtime cost in whatever unit as along as it is the same for all opcodes. My numbers can certainly be improved.
I will add a warning instead of a todo at the beginning of the function to indicate that the numbers are estimates and can be improved.